src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -169,16 +169,53 @@ public class TemplateMatcher
                                          IReadOnlyDictionary<string, object> defaults,
                                          RouteValueDictionary values)
         {
+            var indexOfLastSegment = routeSegment.Parts.Count - 1;
+            if(routeSegment.Parts[indexOfLastSegment].IsOptional && 
+                routeSegment.Parts[indexOfLastSegment - 1].IsOptionalSeperator)
+            {
+                if (MatchComplexSegmentCore(routeSegment, requestSegment, Defaults, values, indexOfLastSegment))
+                {
+                    return true;
+                }
+                else
+                {
+                    if(requestSegment.EndsWith(routeSegment.Parts[indexOfLastSegment-1].Text))
+                    {
+                        return false;
+                    }
+                    
+                    return MatchComplexSegmentCore(
+                        routeSegment, 
+                        requestSegment, 
+                        Defaults, 
+                        values, 
+                        indexOfLastSegment - 2);
+                }
+            }
+            else
+            {
+                return MatchComplexSegmentCore(routeSegment, requestSegment, Defaults, values, indexOfLastSegment);
+            }
+        }
+
+        private bool MatchComplexSegmentCore(TemplateSegment routeSegment,
```

Comment:
Please undo all the changes in this method unless they have a reason to be here.